AnKi/Shaders/VrsSriGeneration.glsl

```diff
@@ -8,57 +8,130 @@
+}
 
 void main()
 {
-	// Get luminance
-	const Vec3 color = texelFetch(u_inputTex, IVec2(gl_GlobalInvocationID.xy), 0).xyz;
-	const F32 luma = computeLuminance(color);
+	IVec2 blockLocation = IVec2(gl_GlobalInvocationID.xy) * IVec2(2, 4);
+
+	// Get luminance.
+	// l0.x  l0.y
+	// l0.z  l0.w  l2.x
+	// l1.x  l1.y
+	// l1.z  l1.w  l2.y
+	//       l2.z
+
+	Vec4 l0;
+	l0.x = SampleLuma(blockLocation + IVec2(0, 0));
+	l0.y = SampleLuma(blockLocation + IVec2(1, 0));
+	l0.z = SampleLuma(blockLocation + IVec2(0, 1));
+	l0.w = SampleLuma(blockLocation + IVec2(1, 1));
+
+	Vec4 l1;
+	l1.x = SampleLuma(blockLocation + IVec2(0, 2));
+	l1.y = SampleLuma(blockLocation + IVec2(1, 2));
+	l1.z = SampleLuma(blockLocation + IVec2(0, 3));
+	l1.w = SampleLuma(blockLocation + IVec2(1, 3));
+
+	Vec3 l2;
+	l2.x = SampleLuma(blockLocation + IVec2(2, 1));
+	l2.y = SampleLuma(blockLocation + IVec2(2, 3));
+	l2.z = SampleLuma(blockLocation + IVec2(1, 4));
+
+	// Calculate derivatives.
+	Vec4 a = Vec4(l0.y, l2.x, l1.y, l2.y);
+	Vec4 b = Vec4(l0.x, l0.w, l1.x, l1.w);
+	Vec4 dx = abs(a - b);
+
+	a = Vec4(l0.z, l0.w, l1.z, l2.z);
+	b = Vec4(l0.x, l0.y, l1.x, l1.w);
+	Vec4 dy = abs(a - b);
+
+	F32 maxDerivativeX = max(max(dx.x, dx.y), max(dx.z, dx.w));
+	F32 maxDerivativeY = max(max(dy.x, dy.y), max(dy.z, dy.w));
+	maxDerivativeX = subgroupMax(maxDerivativeX);
+	maxDerivativeY = subgroupMax(maxDerivativeY);
+
+	// Calculate average luma in block.
+	Vec4 sumL0L1 = l0 + l1;
+	F32 averageLuma = (sumL0L1.x + sumL0L1.y + sumL0L1.z + sumL0L1.w) / 8.0;
+	averageLuma = subgroupAdd(averageLuma);
+
+	// Store results in shared memory.
+	ANKI_BRANCH if(subgroupElect())
+	{
+		s_averageLuma[gl_SubgroupID] = averageLuma;
+		s_maxDerivativeX[gl_SubgroupID] = maxDerivativeX;
+		s_maxDerivativeY[gl_SubgroupID] = maxDerivativeY;
+	}
 
-	// Store luminance
-	s_lumaMin[gl_LocalInvocationIndex] = luma;
-	s_lumaMax[gl_LocalInvocationIndex] = luma;
 	memoryBarrierShared();
 	barrier();
 
-	// Gather the results into one
-	ANKI_LOOP for(U32 s = (WORKGROUP_SIZE.x * WORKGROUP_SIZE.y) / 2u; s > 0u; s >>= 1u)
+	// Write the result
+	ANKI_BRANCH if(gl_LocalInvocationIndex == 0u)
 	{
-		if(gl_LocalInvocationIndex < s)
+		// Get max across all subgroups.
+		averageLuma = s_averageLuma[0];
+		maxDerivativeX = s_maxDerivativeX[0];
+		maxDerivativeY = s_maxDerivativeY[0];
+		ANKI_LOOP for(U32 i = 1u; i < gl_NumSubgroups; ++i)
 		{
-			s_lumaMin[gl_LocalInvocationIndex] =
-				min(s_lumaMin[gl_LocalInvocationIndex], s_lumaMin[gl_LocalInvocationIndex + s]);
-			s_lumaMax[gl_LocalInvocationIndex] =
-				max(s_lumaMax[gl_LocalInvocationIndex], s_lumaMax[gl_LocalInvocationIndex + s]);
+			averageLuma += s_averageLuma[i];
+			maxDerivativeX = max(maxDerivativeX, s_maxDerivativeX[i]);
+			maxDerivativeY = max(maxDerivativeY, s_maxDerivativeY[i]);
 		}
 
-		memoryBarrierShared();
-		barrier();
-	}
+		// Determine shading rate.
+		const Vec2 maxDiff = Vec2(maxDerivativeX, maxDerivativeY);
```

Comment:
Since you are collapsing the max diff should you be carrying the derivatives in both X and Y? Does it make sense to compute a max derivative of both directions in line 80?